### DISCUSSION TITLE
The way to do test

### DISCUSSION THREAD

**Discussion by ZLY201** [js-sdsl/js-sdsl#86] [Ideas]
## Foreword

https://github.com/js-sdsl/js-sdsl/pull/85

We should find a good way to do test.

@noname0310 

## Possible solutions
1. **Add new tests for build mode**
This way will make test files so large.
2. **Make the unit tests done in production mode (after building)**
I haven't found anyone doing like this.
The usual way is to test the source code directly.
And I suggest test for `ts` but not `js`.
3. **Put the build script in a new repository and test it separately**
Such as `sdsl-cli`.
This way will reduce the pressure on the repo.
But we cannot fully guarantee accuracy.
4. **Try to configure the `karma.config.ts` and make it test in building mode**
It's a good idea but I don't know how feasible.

**Comment by noname0310** [comment]
I think the second option is fine.

However, for a complete test, it seems necessary to run the test on the js file.

**Comment by noname0310** [comment]
If only the import could be done dynamically, one test could be run on all builds.

example:
```typescript
import { Vector as CjsVector } from 'dist/cjs/index';
import { Vector as UmdVector } from 'dist/umd/index';
import { Vector as IsolateCjsVector } from 'dist/isolate/cjs/index';
import { Vector as IsolateUmdVector } from 'dist/isolate/umd/index';

let Vector;
if (env.testPackage === 'cjs') {
  Vector = CjsVector;
} else if (env.testPackage === 'umd') {
  Vector = UmdVector;
} else if (env.testPackage === 'isolate-cjs') {
  Vector = IsolateCjsVector;
} else if (env.testPackage === 'isolate-umd') {
  Vector = IsolateUmdVector;
} else {
  throw new Error('unknown package type');
}

describe('SequentialContainer test', () => {
  it('Vector test', () => {
    const myVector = new Vector([1], false);
    //...
  });
});
```

**Comment by ZLY201** [comment]
I need to research the practices of other open source projects to determine the plan.
